// astro/dist/core/errors/dev/vite.js
import * as fs from "fs";
import { fileURLToPath } from "url";
import { createLogger } from "vite";
import { AstroErrorData } from "../errors-data.js";
import { createSafeError } from "../utils.js";
import { incompatPackageExp } from "./utils.js";
function createCustomViteLogger(logLevel) {
  const viteLogger = createLogger(logLevel);
  const logger = {
    ...viteLogger,
    error(msg, options) {
      if (incompatPackageExp.test(msg))
        return;
      return viteLogger.error(msg, options);
    }
  };
  return logger;
}
function enhanceViteSSRError(error, filePath, loader) {
  var _a, _b, _c, _d;
  const safeError = createSafeError(error);
  if (loader) {
    try {
      loader.fixStacktrace(safeError);
    } catch {
    }
  }
  if (filePath) {
    const path = fileURLToPath(filePath);
    const content = fs.readFileSync(path).toString();
    const lns = content.split("\n");
    if (/failed to load module for ssr:/.test(safeError.message)) {
      const importName = (_a = safeError.message.split("for ssr:").at(1)) == null ? void 0 : _a.trim();
      if (importName) {
        safeError.message = AstroErrorData.FailedToLoadModuleSSR.message(importName);
        safeError.hint = AstroErrorData.FailedToLoadModuleSSR.hint;
        safeError.code = AstroErrorData.FailedToLoadModuleSSR.code;
        const line = lns.findIndex((ln) => ln.includes(importName));
        if (line !== -1) {
          const column = (_b = lns[line]) == null ? void 0 : _b.indexOf(importName);
          safeError.loc = {
            file: path,
            line: line + 1,
            column
          };
        }
      }
    }
    if (/Invalid glob/.test(safeError.message)) {
      const globPattern = (_c = safeError.message.match(/glob: "(.+)" \(/)) == null ? void 0 : _c[1];
      if (globPattern) {
        safeError.message = AstroErrorData.InvalidGlob.message(globPattern);
        safeError.hint = AstroErrorData.InvalidGlob.hint;
        safeError.code = AstroErrorData.InvalidGlob.code;
        const line = lns.findIndex((ln) => ln.includes(globPattern));
        if (line !== -1) {
          const column = (_d = lns[line]) == null ? void 0 : _d.indexOf(globPattern);
          safeError.loc = {
            file: path,
            line: line + 1,
            column
          };
        }
      }
    }
  }
  return safeError;
}
function getViteErrorPayload(err) {
  var _a, _b, _c, _d;
  let plugin = err.plugin;
  if (!plugin && err.hint) {
    plugin = "astro";
  }
  const message = `${err.message}

${err.hint ?? ""}`;
  const frame = (_a = err.frame) == null ? void 0 : _a.replace(/\t/g, " ");
  return {
    type: "error",
    err: {
      ...err,
      frame,
      loc: {
        file: (_b = err.loc) == null ? void 0 : _b.file,
        line: ((_c = err.loc) == null ? void 0 : _c.line) ?? 0,
        column: ((_d = err.loc) == null ? void 0 : _d.column) ?? 0
      },
      plugin,
      message: message.trim(),
      stack: err.stack
    }
  };
}
export {
  createCustomViteLogger,
  enhanceViteSSRError,
  getViteErrorPayload
};
